traffic_mirroring.go

```diff
@@ -30,36 +26,65 @@ func main() {
 	}
 
 	packetSource := gopacket.NewPacketSource(handle, handle.LinkType())
-	max_size := 10
-	packets_list := make([][]byte,max_size) 
-	current_packet := 0
+	packetsList := make(chan []byte, 22)
+	filePath := "dump_file"
+	batchSize := 20
 
-	for overlayPacket := range packetSource.Packets() {
-		vxlanLayer := overlayPacket.Layer(layers.LayerTypeVXLAN)
-		if vxlanLayer != nil {
-			vxlanPacket, _ := vxlanLayer.(*layers.VXLAN)
-			
-			fmt.Println(current_packet)
-			packets_list[current_packet] = vxlanPacket.LayerPayload()
-			current_packet += 1
-			fmt.Println(vxlanPacket.LayerPayload())
-			if current_packet == max_size {
-				newSlice := make([][]byte, max_size)
-				copy(newSlice, packets_list)
-				fmt.Println(current_packet)
-				current_packet = 0
-				go write_file(newSlice, f)
+	f, err := os.Create(filePath)
+	check(err)
+	defer f.Close()
 
+	sigChan := make(chan os.Signal)
+	signal.Notify(sigChan, os.Interrupt)
+	go handleSignal(sigChan, packetsList, f, os.Exit)
 
-			}
+	go WriteFile(packetsList, f, batchSize)
+	NetworkListener(packetSource.Packets(), packetsList)
+}
+
+
+func check(e error) {
+	if e != nil {
+		panic(e)
+	}
+}
+
+func handleSignal(sigChan chan os.Signal, packetsList chan []byte, file io.Writer, end func(int)) {
+	<-sigChan
+	exitCode := 130
+	WriteFile(packetsList, file, len(packetsList))
+	end(exitCode)
+}
+
+func NetworkListener(source chan gopacket.Packet, dest chan []byte) {
+	for overlayPacket := range source {
+		vxlanLayer := overlayPacket.Layer(layers.LayerTypeVXLAN)
+		if vxlanLayer == nil {
+			log.Printf("Unable to get VXLAN Layer for packet with metadata (%+v)\n", overlayPacket.Metadata())
+		}
+		vxlanPacket, ok := vxlanLayer.(*layers.VXLAN)
+		if !ok {
+			log.Printf("Unable to cast packet (%+v) to vxlan layer", overlayPacket.Metadata())
 		}
+		dest <- vxlanPacket.LayerPayload()
 	}
-	defer f.Close()
 }
 
-func write_file(packets_list [][]byte, f *os.File) {
-	for _, value := range packets_list {
-		fmt.Fprintln(f, value)  
+func WriteFile(packetsList chan []byte, file io.Writer, batchSize int) {
+	// TODO: on the program exit, you need to write the remain packets inside the channel.
+
+	for {
```

Comment:
Loop until the `packetList` is closed, then write the remaining packets inside the channel.